pkg/steps/source.go

```diff
@@ -539,20 +542,51 @@ func hintsAtInfraReason(logSnippet string) bool {
 		strings.Contains(logSnippet, "connection reset by peer")
 }
 
-func waitForBuildOrTimeout(ctx context.Context, buildClient BuildClient, namespace, name string) error {
-	return waitForBuild(ctx, buildClient, namespace, name)
+func waitForBuildOrTimeout(
+	ctx context.Context,
+	buildClient BuildClient,
+	podClient kubernetes.PodClient,
+	namespace, name string,
+) error {
+	return waitForBuild(ctx, buildClient, podClient, namespace, name)
 }
 
-func waitForBuild(ctx context.Context, buildClient BuildClient, namespace, name string) error {
+func waitForBuild(
```

Comment:
It is as complex as the task it has to perform (and hopefully no more).  I've added both documentation and comments detailing everything that is involved in it, based on the questions here.

It may help to contrast it to [`waitForPodCompletionOrTimeout`](https://github.com/openshift/ci-tools/blob/00cb7dcfe5f92cdf3480208473c99d8316df8657/pkg/util/pods.go#L140), on which it is based (a funny fact, since that one performs a much more complex task, and I seem to have got that one correctly but not this one).